Mainnet WETH Deployment

jflatow opened the pull request:
~~WIP~~ Ready for review now.

Changes made in preparation for launching a ~~mainnet/ETH~~`mainnet/WETH` deployment.

- adds `mainnet-weth` scenario base
- adds configuration for `mainnet/weth` deploy with `wstETH` and `cbETH` as collateral assets
- adds a `WstETHPriceFeed.sol` contract that calculates the `wstETH / USD` price using the `stETH / USD` price feed and the value of `wstETH:stETH`
- updates scenarios to work with ETH-based markets, where possible
- adds a `matchesDeployment` filter to allow for scenarios that specifically target (or avoid) the `mainnet-weth` deploy
- adds a `getRewardToken` fn to context, to allow retrieving the reward token even when it is not a collateral asset

Remaining todos before we can launch:

- [x] set the supply caps back to 0
- [x] replace the `cbETH` price feed with the actual Chainlink `cbETH / USD` price feed (when it's available)
- [x] decide on actual configuration params
- [x] get OZ audit on `WstETHPriceFeed.sol`
- [x] launch a new version of the Liquidator Bot for `mainnet/eth`
- [x] deploy new Bulker (#611)

jflatow committed:
Mainnet ETH Deployment

* Update Github Actions to include mainnet-eth
* ETH-base bulker scenario
* WstETHPriceFeed + tests (#600)
* Non-ETH and ETH bulker scenarios (all actions in one txn)
* Mainnet WETH Bulker (#611)
* Add a supply cap constraint and set initial caps to 0 for cWETHv3
* Update the collateral params based on Gauntlet recommendations (#628)
 https://hackmd.io/wncIvkFTReWUe2AMMK2ezA?view
* Price feeds for WETH deployment + Bulker changes for OZ audit (#625)
 This PR implements and modifies price feeds to support the upcoming WETH deployment. The favored plan so far is to use ETH-denominated price feeds as opposed to USD price feeds, but stick with using 8 decimals for prices to avoid having to change the `Comet` and `Configurator` implementations.
 This would require:
  - A new wrapper price feed (`ScalingPriceFeed.sol`) that scales prices up or down to 8 decimals
  - A new `ConstantPriceFeed` that always returns 1e8 for the `WETH` base asset, since should always hold a 1:1 value with ETH
  - Modifications to the `WstETHPriceFeed` to return prices in terms of ETH instead of USD

 This is an alternative approach to #626, which is a more complex change but could be a better long-term solution.
 *Note: This PR also now contains the changes from #634 and #635, which address some suggestions made by OZ for their audit of `WstETHPriceFeed` and `Bulker`.*

kevincheng96 commented on deployments/mainnet/eth/configuration.json:
```diff
@@ -0,0 +1,49 @@
+    "baseMinForRewards": "1000000e6"
+  },
+  "rewardTokenAddress": "0xc00e94cb662c3520282e6f5717214004a7f26888",
+  "assets": {
```
Thoughts on also supporting a stablecoin like USDC to better facilitate the shorting use-case?

kevincheng96 replied:
Since with the current configuration, the only use-case is levering up on staked ETH.

scott-silver replied:
Sounds like we're sticking with `cbETH` and `wstETH` for the launch

scott-silver commented on .github/workflows/deploy-market.yaml:
```diff
@@ -32,7 +33,7 @@ jobs:
         uses: hayesgm/seacrest@v1
         with:
-          ethereum_url: "${{ fromJSON('{\"fuji\":\"https://api.avax-test.network/ext/bc/C/rpc\",\"kovan\":\"https://kovan.infura.io/v3/$INFURA_KEY\",\"mainnet\":\"https://mainnet.infura.io/v3/$INFURA_KEY\",\"goerli\":\"https://goerli.infura.io/v3/$INFURA_KEY\",\"mumbai\":\"https://polygon-mumbai.infura.io/v3/$INFURA_KEY\"}')[inputs.network] }}"
+          ethereum_url: "${{ fromJSON('{\"fuji\":\"https://api.avax-test.network/ext/bc/C/rpc\",\"kovan\":\"https://kovan.infura.io/v3/$INFURA_KEY\",\"mainnet\":\"https://mainnet.infura.io/v3/$INFURA_KEY\",\"mainnet-eth\":\"https://mainnet.infura.io/v3/$INFURA_KEY\",\"goerli\":\"https://goerli.infura.io/v3/$INFURA_KEY\",\"mumbai\":\"https://polygon-mumbai.infura.io/v3/$INFURA_KEY\"}')[inputs.netw ...
```
added the `mainnet-eth` deployment to the Github Actions.

for the `deploy-market` action, `mainnet` uses `https://mainnet.infura.io/v3/$INFURA_KEY`, but for `prepare-migration` and `enact-migration` it uses `https://mainnet-eth.compound.finance\`

wondering if there is a reason for the difference? or should we pick one and use them across the actions

scott-silver replied:
updated

scott-silver commented on deployments/mainnet/eth/configuration.json:
```diff
@@ -0,0 +1,49 @@
+      "borrowCF": 0.80,
+      "liquidateCF": 0.85,
+      "liquidationFactor": 0.95,
+      "supplyCap": "500e18"
```
bumped the supply caps to get scenarios running

- [ ] reset supplyCaps to 0

Cumop21 commented on .github/workflows/deploy-market.yaml:
```diff
@@ -32,7 +33,7 @@ jobs:
       - name: Seacrest
         uses: hayesgm/seacrest@v1
         with:
-          ethereum_url: "${{ fromJSON('{\"fuji\":\"https://api.avax-test.network/ext/bc/C/rpc\",\"kovan\":\"https://kovan.infura.io/v3/$INFURA_KEY\",\"mainnet\":\"https://mainnet.infura.io/v3/$INFURA_KEY\",\"goerli\":\"https://goerli.infura.io/v3/$INFURA_KEY\",\"mumbai\":\"https://polygon-mumbai.infura.io/v3/$INFURA_KEY\"}')[inputs.network] }}"
```
jflatow/mainnet-eth

scott-silver commented on scenario/SupplyScenario.ts:
```diff
@@ -250,7 +250,7 @@ scenario(
         amount: 100n * scale,
       })
-    ).to.be.revertedWith('ERC20: transfer amount exceeds allowance');
+    ).to.be.reverted;
```
`WETH` reverts without reason strings, unfortunately ☹️

jflatow replied:
This kinda sucks, another case where I think we should at least preserve the error information in a comment in case we find a way to restore strength later

scott-silver replied:
sure, added the error messages back as comments.

as long as we're running these scenarios against a WETH-based deployment we'll have to stick with `.reverted`, though, since the WETH contract we're using isn't ever going to return revert strings.

jflatow replied:
maybe one day we'll have network specific revert reasons 🤷

scott-silver replied:
oh, that's an interesting idea.

we could write a wrapper function that the function to execute and a map of networks to the expected error message. for anything

scott-silver commented on scenario/RewardsScenario.ts:
```diff
@@ -156,8 +156,8 @@ scenario(
-    const [rewardTokenAddress, rescaleFactor] = await rewards.rewardConfig(comet.address);
-    const rewardToken = context.getAssetByAddress(rewardTokenAddress);
+    const [_, rescaleFactor] = await rewards.rewardConfig(comet.address);
+    const rewardToken = await context.getRewardToken();
```
we could have `context.getRewardToken` return both the `rewardToken` instance and the `rescaleFactor`, which would eliminate the need to call `rewards.rewardConfig` on line 159

jflatow replied:
`getRewardConfig`?

scott-silver replied:
i like it. added a `getRewardConfig` function context.

jflatow commented on .github/workflows/deploy-market.yaml:
```diff
@@ -9,6 +9,7 @@ on:
           - fuji
           - kovan
           - mainnet
+          - mainnet-eth
```
Can we get rid of this now?

scott-silver replied:
yep, removed

jflatow commented on deployments/mainnet/eth/deploy.ts:
```diff
@@ -0,0 +1,45 @@
+import { getConfiguration, NetworkConfiguration } from '../../../src/deploy/NetworkConfiguration';
+
+export default async function deploy(deploymentManager: DeploymentManager, deploySpec: DeploySpec): Promise<Deployed> {
+  // XXX configuration params for assets are just placeholders
```
We should ask Gauntlet to start thinking about params too

scott-silver replied:
Started a conversation in the Gauntlet channel

jflatow commented on deployments/mainnet/eth/deploy.ts:
```diff
@@ -0,0 +1,45 @@
+  const { comet } = deployed;
+
+  // Get a handle to the WETH contract for bulker
+  const WETH = await deploymentManager.existing('WETH', await comet.baseToken());
```
Is it not in spider yet? (e.g. `deploymentManager.contract('WETH')`)

scott-silver replied:
we don't even need to do that. we can just pass in `await comet.baseToken()` to the Bulker constructor. updated

jflatow commented on deployments/mainnet/eth/deploy.ts:
```diff
@@ -0,0 +1,45 @@
+  const deployed = await deployComet(
+    deploymentManager,
+    deploySpec,
+    { assetConfigs }
```
Something seems off here, we shouldn't have to do all this... 🤔

scott-silver replied:
The additional setup is so that we can deploy the `wstETHPriceFeed` contract and then pass it in as an argument to `deployComet`, overriding the value from `configuration.json`.

We normally don't have to do this since all of our price feeds are already deployed.

One other option would be to deploy the price feed manually in advance, and just add that address to the configuration file.

jflatow replied:
I think we might be able to get rid of everything except the deploy if we use ```getContractAddress(`${assetName}:priceFeed`, assetConfig.priceFeed)``` here https://github.com/compound-finance/comet/blob/main/src/deploy/NetworkConfiguration.ts#L98 i.e. with the current one as the fallback (and I guess here too for parity https://github.com/compound-finance/comet/blob/main/src/deploy/NetworkConfiguration.ts#L134)

scott-silver replied:
that's awesome. made the update

jflatow reviewed (commented):
Looking great, couple comments/questions 💪

jflatow commented on deployments/mainnet/eth/deploy.ts:
```diff
@@ -0,0 +1,45 @@
+
+  // Deploy WstETHPriceFeed
+  const wstETHPriceFeed = await deploymentManager.deploy(
+    'wstETHPriceFeed',
```
The alias should be the same as what spider gets, which I think should be `wstETH:priceFeed`

jflatow commented on deployments/mainnet/eth/deploy.ts:
```diff
@@ -5,28 +5,16 @@ import { getConfiguration, NetworkConfiguration } from '../../../src/deploy/Netw
   // Deploy WstETHPriceFeed
   const wstETHPriceFeed = await deploymentManager.deploy(
-    'wstETHPriceFeed',
+    'wstETH:priceFeed',
```
🎉

jflatow commented on deployments/mainnet/eth/configuration.json:
```diff
@@ -0,0 +1,49 @@
+  "assets": {
+    "cbETH": {
+      "address": "0xBe9895146f7AF43049ca1c1AE358B0541Ea49704",
+      "priceFeed": "0xCfE54B5cD566aB89272946F602D76Ea879CAb4a8",
```
XXX change to upcoming cbETH/USD feed

jflatow replied:
Updated to https://etherscan.io/address/0x67eF3CAF8BeB93149F48e8d20920BEC9b4320510

jflatow commented:
For setting the supply caps from 0, we need a separate migration w/ a proposal. For USDC we also seeded reserves, not sure if we want to try to do that here. Doing rewards after I think makes sense

jflatow commented on scenario/BulkerScenario.ts:
```diff
@@ -168,3 +169,94 @@ scenario(
+    expect(await baseAsset.balanceOf(albert.address)).to.be.equal(toBorrowBase);
+    expect(await comet.balanceOf(betty.address)).to.be.equal(baseTransferred);
+    // NOTE: differs from the equivalent scenario for non-ETH markets
+    expect(await comet.borrowBalanceOf(albert.address)).to.be.equal(toBorrowBase + toTransferBase - (toSupplyEth - toWithdrawEth));
```
Just to make sure I understand, if the scenario changed the collateral asset from ETH to something else it would be the same right?

scott-silver replied:
The `non-WETH` version of this test has these two assertions instead:

```
expect(await comet.borrowBalanceOf(albert.address)).to.be.equal(toBorrowBase + toTransferBase);
expect(await comet.collateralBalanceOf(albert.address, WETH.address)).to.be.equal(toSupplyEth - toWithdrawEth);
```

So in a USDC deployment, the borrow balance is how much the user borrowed and transferred. As well, the user has a collateral balance of WETH equal to how much ETH they supplied minus how much they borrowed.

But in this case, since the base is WETH, then the borrow balance is how much WETH was borrowed, how much WETH was transferred, how much ETH was borrowed (minus how much ETH was supplied)

jflatow commented:
- [ ] We also need to make the name consistent everywhere (cWETHv3)

kevincheng96 commented on scenario/GovernanceScenario.ts:
```diff
@@ -143,7 +143,68 @@ scenario('add new asset',
+    expect(await albert.getCometBaseBalance()).to.be.equal(-borrowAmount);
+  });
+
+scenario('add new asset (mainnet-weth)',
```
Curious why we need to separate out this scenario? Shouldn't it just work for both deployments?

scott-silver replied:
the original test attempts to borrow 1,000 units of the base asset using 100 assets of DOGE (with a price of $1,000 per DOGE).

that works with a USDC base, since a $1,000 borrow can be collateralized with $100,000 worth of DOGE.

but borrowing 1,000 ETH costs about $1.5M, so the borrow is uncollateralized

kevincheng96 replied:
Ah I see, that is indeed annoying. Another approach could be to set the borrow amount to be the `borrowBaseMin` instead of a constant. But this works too.

jflatow replied:
Shouldn't we just have a PriceConstraint?

scott-silver replied:
that's a clever idea. So you'd specify a price for the base asset and/or the collateral assets

we're having a similar conversation on the [Polygon PR](https://github.com/compound-finance/comet/pull/598#discussion_r1013496694).

Another approach would be to write a helper function returned the amount of a given asset required to cover a borrow of X base units

jflatow replied:
Haha, yep just commented there, we can pick one ;) But I think its more desirable to use literal numbers in properties where we can (👨‍🍳 has long advocated for this and I kind of agree)

scott-silver replied:
#615

kevincheng96 commented on scenario/RewardsScenario.ts:
```diff
@@ -21,7 +21,7 @@ function calculateRewardsOwed(
   'Comet#rewards > can claim supply rewards for self',
   {
-    filter: async (ctx) => await isRewardSupported(ctx),
+    filter: async (ctx) => await isRewardSupported(ctx) && !matchesDeployment(ctx, [{network: 'mainnet', deployment: 'weth'}]),
```
We should add an XXX to remove this filter once rewards are supported on the WETH deployment.

scott-silver replied:
i guess alternatively, we could delete `rewardTokenAddress` from the `mainnet/WETH` configuration so that the `isRewardSupported` check works as intended

scott-silver replied:
deleted `rewardTokenAddress` from the configuration file, and deleted this filter

kevincheng96 commented on test/wsteth-price-feed.ts:
```diff
@@ -0,0 +1,107 @@
+  };
+}
+
+const testCases = [
```
Weird edge case to maybe test is if the price returned is negative (Chainlink returns `int256` for price). I don't think it really matters since Comet will just revert if it's a negative price, but let's make sure we do return a negative price back.

scott-silver replied:
good idea. added a negative price to the test cases

kevincheng96 reviewed (commented):
Looks great so far!

scott-silver commented on contracts/WstETHPriceFeed.sol:
```diff
@@ -0,0 +1,76 @@
+    ) {
+        (uint80 roundId, int256 stETHPrice, uint256 startedAt, uint256 updatedAt, uint80 answeredInRound) = AggregatorV3Interface(stETHtoUSDPriceFeed).latestRoundData();
+        uint256 tokensPerStEth = IWstETH(wstETH).tokensPerStEth();
+        int price = stETHPrice * signed256(wstETHScale) / signed256(tokensPerStEth);
```
yep, that works too. updated

kevincheng96 reviewed (commented):
Cool, I think everything looks good now. Just waiting on params to be finalized

jflatow commented on contracts/WstETHPriceFeed.sol:
```diff
@@ -70,7 +70,7 @@ contract WstETHPriceFeed is AggregatorV3Interface {
         (uint80 roundId, int256 stETHPrice, uint256 startedAt, uint256 updatedAt, uint80 answeredInRound) = AggregatorV3Interface(stETHtoUSDPriceFeed).latestRoundData();
         uint256 tokensPerStEth = IWstETH(wstETH).tokensPerStEth();
-        int price = stETHPrice * signed256(wstETHScale) / signed256(tokensPerStEth);
+        int256 price = stETHPrice * int256(wstETHScale) / signed256(tokensPerStEth);
```
Since we are assuming `wstETHScale` is already guaranteed to be < int256.max?

scott-silver replied:
yeah, that's the assumption:

https://github.com/compound-finance/comet/pull/587#discussion_r1009744397

jflatow replied:
Yep 👍 comet has `10**18` max constant and changing that would be a massive breaking change

kevincheng96 commented on scenario/constraints/SupplyCapConstraint.ts:
```diff
@@ -0,0 +1,60 @@
+          const asset = await getAssetFromName(assetName, context);
+          const decimals = await asset.token.decimals();
+          const amount = parseAmount(rawAmount);
+          expect(amount.op).to.equal(ComparisonOp.GTE, `Operation ${amount.op} not supported (yet) by supply cap constraint`);
```
What does this line do?

scott-silver commented on deployments/mainnet/weth/configuration.json:
```diff
@@ -0,0 +1,48 @@
+  "baseToken": "WETH",
+  "baseTokenAddress": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
+  "baseTokenPriceFeed": "0x5f4eC3Df9cbd43714FE2740f5E3616155c5b8419",
+  "borrowMin": "100e6",
```
bumping this so we don't forget: `borrowMin`, `targetReserves`, and `baseMinForRewards` are all 6 decimal values, not 18 decimal values

kevincheng96 replied:
Here's what I'm thinking for the tracking index values:
```
trackingIndexScale: 1e12 or 1e15
baseMinForRewards: 1_000e18 (is this too high? similar to USDC market, this would be around $1mn)
```

jflatow committed:
Refactor deploy to support second markets (#639)

* Refactor deploy to support second markets

* Add a fromDep to deployment manager which can be used for sharing dependency contracts from another deployment
* Always attempt to initialize storage if needed, and not necessarily as gov
* Always deploy an initial implementation contract instead of using the factory marker
* Only take gov admin actions if we own the cometAdmin
* Expose the primitives necessary for a proposal to configurate later, if/when admin cannot

* Add an env flag for migration constraint to skip non-migration scenarios
* Defer reading current config in modern constraint (bugfix)

* Add recipient to token sourcing blacklist

kevincheng96 committed:
WETH deployment audit changes (#641)

These are changes in response to the WETH deployment [audit](https://gist.github.com/andresbach/a01686b1ef8bdf51d46046f2cfc93307) conducted by OZ:

- **L01 - Lack of input validation** - We originally avoided zero address checks because there are a vast amount of accidental addresses that can be set here and checking for a specific one seems unnecessary. However, we have reconsidered that position after seeing the reasons for including a zero address check listed in this [post](https://forum.openzeppelin.com/t/removing-address-0x0-checks-from-openzeppelin-contracts/2222/13).
- **L02 - Improper implementation of Chainlink AggregatorV3Interface** - We updated the hard-coded return values for `roundId` and `answeredInRound` to be 1 to comply with the Chainlink specification.
- **L04 - Incomplete/confusing documentation** - We updated the documentation based on the provided suggestions.

Our full audit responses are [here](https://docs.google.com/document/d/1A8TrUdTtD7DXr1U5HOqSS1Xl81rjBZdlDC7LS4No4Bw).

kevincheng96 committed:
Spider USDC deployment before calling `fromDep` (#644)

We need to spider the `USDC` deployment before calling `fromDep` on it so all the aliases are populated in the cache. Otherwise, `fromDep` may try to import an `undefined` contract. 

There are many approaches to this and this PR takes the simplest approach of spidering in the deploy script. Another approach could be to handle the spidering directly in `fromDep`.

kevincheng96 committed:
Convert from scientific notation to bigint in configuration (#643)

This PR adds a new `ScientificNotation` string type to `NetworkConfiguration` that is required for certain fields (e.g. `supplyCap`, `baseMinForRewards`). The configuration parser will safely convert these values into bigints without losing precision; with one caveat, scientific notations that use decimals in the coefficient still run the risk of precision loss because those are converted directly to javascript numbers.

jflatow committed:
Explicit supply cap constraint for failing bulker scen (#648)

kevincheng96 committed:
Add tracking index bounds unit tests for WETH (#645)

jflatow committed:
Update params and rate model final numbers for deployment

scott-silver committed:
OnChainLiquidator (#642)

Update the reference liquidator to source from more pools and generally be more robust

jflatow committed:
Fix fromDep to spider the dependency and load the contract from it (#652)

scott-silver committed:
bump targetReserves in LiquidationBotScenarios (#653)

scott-silver committed:
delete CURVE_REGISTRY (#654)

scott-silver committed:
OnChainLiquidator, not Liquidator (#655)

jflatow committed:
Combined model based on Gauntlet's first and second recommendations (#656)

scott-silver committed:
Wrap and unwrap native token as base asset in Bulker (#659)

* Add tests for supply native token; clean up
* Change supplyStETH to use stETH amount instead of wstETH amount; add scenarios
* Tweak comments
* Supplying uint256.max collateral reverts in WETH deposit not Comet

Commit:
Modified deployment roots from GitHub Actions

jflatow committed:
Proposal to enable cWETHv3 market (#608)

* Proposal to enable cWETHv3 market

* Modify proposal to add rewards
* Update the supply caps based on Gauntlet recommendations (#629)
 https://hackmd.io/wncIvkFTReWUe2AMMK2ezA?view
* Refactor proposal to make market live (#640)

* If there are migrations, don't run the non-migrations

This allows a proposal branch to pass CI if the migration is required and is generally the intention for migration branches.

However an env var is still supported to override and force running the non-migration, too.

* Final changes to supply caps; proposal text; WETH transfer amt (#662)

Also copy bulker to usdc roots and fix scen test fixes

* Deploy cWETHv3 to Goerli (#646)

This PR adds a new `goerli-weth` base and deploy script for the Goerli cWETHv3 deploy.

The deploy script is very similar to the mainnet cWETHv3 script. The main differences are:
- Deploying price feeds for stETH / ETH and cbETH / ETH
- Pulling in extra testnet/cross-chain contracts such as `fauceteer` and `fxRoot` using `fromDep`
- Cloning over `cbETH` from mainnet
- Minting `cbETH` to `fauceteer`

* Modified migration from GitHub Actions

Co-authored-by: Kevin Cheng <kevincheng96@hotmail.com>
Co-authored-by: GitHub Actions Bot <>
